### ISSUE TITLE
Migrate to using Statement library

### ISSUE THREAD

**Issue by Micalobia** [DomiIRL/FixedWaterlogging#6]
You're adding blockstates without Statement, which is generally unsafe and can cause crashes in datapacks that add structures using blocks you've changed, as well as the every block turning into a different block mentioned in #4 

It's not terrible to add blockstates, if you need somewhere to look you can look at my mod [Full Slabs](https://github.com/Micalobia/Full-Slabs). I used to not use Statement and it ended up biting me later, it makes life simpler.

**Comment by DomiIRL** [comment]
Hey there, would you be able to review https://github.com/KxmischesDomi/FixedWaterlogging/commit/8cc6fe3509d3e7fee42c1a75d374bbd5c9323382 so I'm not doing anything wrong and messing up everything

**Comment by supersaiyansubtlety** [comment]
I was just about to suggest using statement as well 😆 

I see a potential reason your `onInitialize` method wasn't initializing states properly:

[here](https://github.com/KxmischesDomi/FixedWaterlogging/blob/statement/src/main/java/de/kxmischesdomi/fixedwaterlogging/FixedWaterloggingMod.java#L19) your `injectBlockProperty` call doesn't need to be inside and event, in fact it may cause issues to delay it until the event fires. You can just call `StateRefresher.INSTANCE.addBlockProperty` for each class you're adding waterlogging to.

I don't think you'll have to call `StateRefresher.INSTANCE.addBlockProperty` in any mixins if the above fixes your initializer.

One other thing I noticed: 
[here](https://github.com/KxmischesDomi/FixedWaterlogging/blob/statement/src/main/java/de/kxmischesdomi/fixedwaterlogging/FixedWaterloggingMod.java#L33) it looks like you could use a simple `instanceof` check rather than reflection.

**Comment by DomiIRL** [comment]
I changed everything as you said and it seems to work fine now. 
But When I first loaded the world I was previously testing in it seemed like all the old blocks that didn't have the waterlogged property until now had it on true.
Newly generated blocks weren't affected tho.
Any ideas on that @supersaiyansubtlety?

**Comment by supersaiyansubtlety** [comment]
Yeah if you make changes to blockstates, especially removing properties (which is effectively what happens when you change the properties you add: remove one and add another), blocks already in-world with the old states can become unexpected states or even other blocks. 

I wouldn't worry about it, just let users know they may have to re-place/re-waterlog any affected blocks in-world.

**Comment by DomiIRL** [comment]
So you think it's alright as it is now?

**Comment by supersaiyansubtlety** [comment]
I haven't downloaded and tested, but it looks right; and based on your description I don't think there's a problem.

**Comment by DomiIRL** [comment]
I love it, can't start the game outside the dev enviroment and the logs giving no information why

**Comment by supersaiyansubtlety** [comment]
put full log on a paste site?

**Comment by TelepathicGrunt** [comment]
Just linking bug report to my mod where current FixedWaterlogging is crashing my mod's block that extends VineBlock: https://github.com/TelepathicGrunt/Bumblezone-Fabric/issues/42

I'll look into trying to move away from using vine block (i was using to to reduce code duplication). But I want to see if you moving to Statement would fix the issue automatically as then I won't need to do a workaround to avoid your mod's mixins from breaking my blocks.

> I love it, can't start the game outside the dev enviroment and the logs giving no information why

Try sharing your latest.log file after unable to start the game. Maybe there's info there about the issue

**Comment by DomiIRL** [comment]
you can just try it yourself by downloading the last build https://github.com/KxmischesDomi/FixedWaterlogging/actions/runs/1655433912

**Comment by TelepathicGrunt** [comment]
looks like it is working for me. No crash and can start game
![image](https://user-images.githubusercontent.com/40846040/149627224-a8b45e43-0b10-4295-b368-7317132712e3.png)
